test/libraries/encoders/TokenRegistry.sol

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+pragma solidity ^0.8.26;
+
+import {IERC20, GPv2Order} from "src/contracts/libraries/GPv2Order.sol";
+
+type Registry is bytes32;
+
+library TokenRegistry {
+    struct State {
+        IERC20[] tokens;
+        mapping(IERC20 => uint256) tokenIndices;
+        mapping(IERC20 => uint256) prices;
+    }
+
+    bytes32 internal constant STATE_STORAGE_SLOT = keccak256("TokenRegistry.storage");
+
+    /// @dev We want to make sure that the token registry never allocates
+    /// the first `tokens` entry to a valid token.  This is because in the
+    /// inverse token mapping in storage we can't distinguish the cases
+    /// "mapping has no entry" and "mapping has entry zero".
+    modifier hydrateArray(State storage state) {
+        if (state.tokens.length == 0) {
+            state.tokens.push(IERC20(address(uint160(uint256(keccak256("TokenRegistry: invalid token placeholder"))))));
+        }
+        _;
+    }
+
+    /// @dev Allocates a new token registry for the specified registry ID
+    function tokenRegistry(Registry id) internal pure returns (State storage state) {
+        bytes32 slot = keccak256(abi.encodePacked(STATE_STORAGE_SLOT, id));
+        assembly {
+            state.slot := slot
+        }
+    }
+
+    /// @dev Retrieve the token index for the specified token address. If the token
+    /// is not already in the registry, it will be added.
+    function indexOf(State storage state, IERC20 token) internal hydrateArray(state) returns (uint256 i) {
```

Comment:
I'd suggest using a different name for this function, `indexOf` is very  misleading because the name hides the side effects. Maybe:

```suggestion
    function pushfNotPresentIndexOf(State storage state, IERC20 token) internal hydrateArray(state) returns (uint256 i) {
```

Motivation is that blocks like this are very perplexing to me:
```
        for (uint256 i = 0; i < _tokens.length; i++) {
            IERC20 token = _tokens[i];
            indexOf(state, token); // unnecessary code?
            state.prices[token] = _prices[i];
        }
```

Reply:
Done